React example (using Vue example as reference)

themetalfleece opened the pull request:
Hello from GitHub too 👋 

I'm gonna build my app using React, and I can see that there is no React example (and there's a stale PR for it).
Since I copied the Vue example into a fresh React project anyway, I was curious if you're interested in adding it as a React example.
It's basically the Vue example with setting up the worker (using `comlink` etc) and displaying the first page as an image. It just uses React hooks to provide the needed functionality.
It was bootstrapped with Vite and uses typescript.

Let me know if it's something you'd like in your codebase so I can properly update files like the Readme and the icon, and polish it a bit more.

TODOs:

- [x] Enable production builds

![image](https://github.com/user-attachments/assets/207a8ec2-84f6-457b-ae8f-b5348c0b2898)

themetalfleece commented:
I have read the CLA Document and I hereby sign the CLA

themetalfleece commented:
recheck

themetalfleece commented:
@jamie-lemon added a basic README, also added a TODO list in the PR's description. Looks like I don't have permissions to add you as a reviewer, but this is ready for review and testing :) (everything works as intended for me)

jamie-lemon commented on examples/react/README.md:
```diff
@@ -0,0 +1,24 @@
+# MuPDF with React, TypeScript, Vite
+
+## Notable files
+- `src/hooks/mupdf.worker.ts`: imports the worker script, exposes functionality from it.
```
`src/hooks/mupdf.worker.ts` -> `src/workers/mupdf.worker.ts`

jamie-lemon commented on examples/react/src/workers/mupdf.worker.ts:
```diff
@@ -0,0 +1,63 @@
+/// <reference lib="webworker" />
+import * as Comlink from "comlink";
+import { ColorSpace, Document } from "mupdf";
```
We can also use the higher level wrapper ( which is preferable ) 
So:
`import { ColorSpace, Document } from "mupdf";` - > `import { ColorSpace, PDFDocument } from "mupdf/mupdfjs";`

jamie-lemon commented on examples/react/src/workers/mupdf.worker.ts:
```diff
@@ -0,0 +1,63 @@
+const mupdfScript = import.meta.env.PROD
+  ? "/assets/mupdf.js"
+  : "/node_modules/mupdf/dist/mupdf.js";
+
```
If we are going ahead with the high level wrapper then target them:
```
? "/assets/mupdfjs.js"
  : "/node_modules/mupdf/dist/mupdfjs.js";
  ```

jamie-lemon commented on examples/react/src/workers/mupdf.worker.ts:
```diff
@@ -0,0 +1,63 @@
+  : "/node_modules/mupdf/dist/mupdf.js";
+
+export class MupdfWorker {
+  private mupdf?: {
```
Again, if going with higher level wrapper we might want to change this instance variable name to `mupdfjs` just for our own sanity.

jamie-lemon commented on examples/react/src/workers/mupdf.worker.ts:
```diff
@@ -0,0 +1,63 @@
+    ColorSpace: typeof ColorSpace;
+  };
+  private document?: Document;
+
```
If going with higher level wrapper, then : `Document` - >`PDFDocument`

jamie-lemon commented on examples/react/src/workers/mupdf.worker.ts:
```diff
@@ -0,0 +1,63 @@
+  async loadDocument(document: ArrayBuffer): Promise<boolean> {
+    if (!this.mupdf) throw new Error("MuPDF not initialized");
+
+    this.document = this.mupdf.Document.openDocument(
```
if higher level wrapper ( and we renamed the `mupdf` var then) :

`this.document = this.mupdfjs.PDFDocument.openDocument(`

jamie-lemon reviewed (changes requested):
This works great! Thanks for taking the time to do this :)

I noticed one typo in the README and aside for that I added some ideas about how to target the higher level wrapper for MuPDF.js ( namely mupdfjs.js which sits onto of mupdf.js ) . The higher level wrapper aims to be a lot more user friendly and to, well "wrap" more things. Why not upgrade your react example to use it?

themetalfleece commented:
> I added some ideas about how to target the higher level wrapper for MuPDF.js ( namely mupdfjs.js which sits onto of mupdf.js ) . The higher level wrapper aims to be a lot more user friendly and to, well "wrap" more things.

So my issue with this is that it starts with `import * as mupdf from "mupdf";` which doesn't work when loading the script on runtime (for prod builds).

However, I adjusted how the worker is loaded, and it now seems to work fine even for production builds (I deviated from the route of the vue example). I'm loading the module directly, and have set-up vite to exclude it from pre-bundling (check out `vite.config.mjs`).
We also now have optimal typing as it's imported directly.

There are a few warnings on compilation time that I feel are safe to ignore, but no warning/errors on runtime.

I pushed the changes, please perform a fresh review and testing session for both dev and prod 🙏

jamie-lemon reviewed (approved):
All good.
